### PR TITLE
Working on streamlining the Maven build

### PR DESCRIPTION
**Pull request by chrisdutz**
- Added support for MavenWrapper
- Refactored the build to always generate thrift sources without the need to install an executable
- Fixed most issues found in the build
- Aligned the build with how Apache expects incubating projects to be built
- Added some recommended additional tooling

### CODE REVIEW

File: grafana/pom.xml [hunk 1]
```diff
@@ -2,73 +2,61 @@
...
+		<artifactId>root</artifactId>
+		<version>0.8.0-SNAPSHOT</version>
+		<relativePath>../pom.xml</relativePath>
+	</parent>
+	<!--parent>
 		<groupId>org.springframework.boot</groupId>
 		<artifactId>spring-boot-starter-parent</artifactId>
 		<version>1.5.4.RELEASE</version>
-		<relativePath /> <!-- lookup parent from repository -->
-	</parent>
```
Comment (MyXOF) [inline grafana/pom.xml]: This dependency is a little bit special. If you remove springframework dependency here, then the war file can't be executed correctly. Becuase this springframework add some dependencies and plugins from parent pom file. 
If you use Eclipse, you can check the effective pom.
So I suggest that grafana project should be considered as a single private project and only provide executable war file to users. not open-source.
@jixuan1989

**Comment by jixuan1989** [comment]
Great!

My mates, I suggest that we take half of time to learn how Chris uses Maven carefully by reviewing these codes.

File: grafana/pom.xml (hunk 1 above)
Comment (chrisdutz) [reply grafana/pom.xml]: In that case we would have to add what's missing ... I'm working on it.

Comment (jixuan1989) [reply grafana/pom.xml]: There is no conflict that making Grafana-connector as an open source project.

**Comment by jixuan1989** [comment]
This PR is for fixing  issue  #500   and #501

File: grafana/pom.xml (hunk 1 above)
Comment (MyXOF) [reply grafana/pom.xml]: It is a little bit challenging, springframework has may dependencies and plugins which are difficult to decide whether to add them to our project or not.
After you finish adding dependencies and plugins , run “mvn clean package” and get war file
First you need to copy application.properties in "conf" directory to target directory , then run "java -jar iotdb-grafana-0.8.0-SNAPSHOT.war"

If you see the below picture, then the work is done.
image

Comment (chrisdutz) [reply grafana/pom.xml]: Well the main problem is the dependency versions and that can easily be solved by importing that information ... as I mentioned ... I'm still working on this particular part of the build.

File: pom.xml [hunk 2]
```diff
@@ -31,43 +36,33 @@
...
+		<maven.compiler.target>1.8</maven.compiler.target>
+
+		<maven.assembly.version>2.5.5</maven.assembly.version>
+		<scala.version>2.0.1</scala.version>
+		<hadoop.version>2.7.3</hadoop.version>
+		<junit.version>4.12</junit.version>
+		<slf4j.version>1.7.12</slf4j.version>
+		<logback.version>1.1.3</logback.version>
+		<json.version>20140107</json.version>
+		<joda.version>2.9.9</joda.version>
```
Comment (jixuan1989) [inline pom.xml]: As mentioned in the mail list: 
> You are relying on Java 8 so the Jodatime library is included in general there is no need to use the external library and the included classes could be used)

I check the source code in IoTDB, it seems that only spring-boot use joda. @MyXOF  can you double check it again?

Comment (MyXOF) [reply pom.xml]: As far as I can see, jdbc, cli and iotdb also use  joda

Comment (chrisdutz) [reply pom.xml]: If you do a full-text search on "import org.joda.time" you'll find a lot more ...

File: grafana/pom.xml (hunk 1 above)
Comment (chrisdutz) [reply grafana/pom.xml]: Ok ... so with my latest commit, the spring boot application seems to be runnable.

Comment (MyXOF) [reply grafana/pom.xml]: I use your grafana/pom.xml and run in Win10, then i got :
`
$ java -jar iotdb-grafana-0.8.0-SNAPSHOT.war
Failed to instantiate [ch.qos.logback.classic.LoggerContext]
Reported exception:
java.lang.AbstractMethodError: ch.qos.logback.classic.pattern.EnsureExceptionHandling.process(Lch/qos/logback/core/Context;Lch/qos/logback/core/pattern/Converter;)V
        at ch.qos.logback.core.pattern.PatternLayoutBase.start(PatternLayoutBase.java:89)
        at ch.qos.logback.classic.encoder.PatternLayoutEncoder.start(PatternLayoutEncoder.java:28)
        at ch.qos.logback.classic.BasicConfigurator.configure(BasicConfigurator.java:50)
        at ch.qos.logback.classic.util.ContextInitializer.autoConfig(ContextInitializer.java:164)
        at org.slf4j.impl.StaticLoggerBinder.init(StaticLoggerBinder.java:85)
        at org.slf4j.impl.StaticLoggerBinder.<clinit>(StaticLoggerBinder.java:55)
        at org.slf4j.LoggerFactory.bind(LoggerFactory.java:150)
        at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:124)
        at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:412)
        at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:357)
        at org.apache.commons.logging.impl.SLF4JLogFactory.getInstance(SLF4JLogFactory.java:155)
        at org.apache.commons.logging.impl.SLF4JLogFactory.getInstance(SLF4JLogFactory.java:132)
        at org.apache.commons.logging.LogFactory.getLog(LogFactory.java:273)
        at org.springframework.boot.SpringApplication.<clinit>(SpringApplication.java:179)
        at cn.edu.tsinghua.web.TsfileWebDemoApplication.main(TsfileWebDemoApplication.java:10)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
        at org.springframework.boot.loader.WarLauncher.main(WarLauncher.java:59)
`

Add the following dependency in <dependencyManagement> can fix this problem and run successfully.
image

Comment (chrisdutz) [reply grafana/pom.xml]: Didn't I add exactly that dependency? Did you rebuild the entire project? Because I bumped the version in the root pom.

Comment (MyXOF) [reply grafana/pom.xml]: I tried again with root pom, it works. Everything is all right now

**Pull request merged**